doc/release/upcoming_changes/26081.improvement.rst

```diff
@@ -0,0 +1,41 @@
+``ndarray`` Shapes are now covariant and bound
+---------------------------------------------------------------
+Static typing for ``ndarray`` is a long-term effort that continues 
+with this change.  Arrays are a generic type, parameterized by both 
+shape and data type.  The previous shape-type placeholder could previously be assigned
+any value.  This change restricts it to a tuple of ints, as one would expect
+from using ``ndarray.shape``.  However, this change also allows parameterizing
+shape by a tuple of ``int`` subtypes, which are reflected in the type of
+``ndarray.shape``, and allows functions to statically manipulate axes:
+
+.. code::
+
+   from typing import Any, NewType, TypeVar
+   import numpy as np
+
+   Time = NewType("Time", int)
+   Series = NewType("Series", int)
+   DType = TypeVar("DType", bound=np.dtype)
+   Ax1 = TypeVar("Ax1", bound=int)
+   Ax2 = TypeVar("Ax2", bound=int)
+
+   arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```

Comment:
If I understand correctly, type declarations are valid so long as they narrow the inferred type.  Since no array creation routines infer the `_ShapeType` from it's arguments,
```
arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```
is valid, as is 
```
arr: np.ndarray[tuple[Literal[2], Literal[2]], Any] = np.arange(4.0).reshape((2, 2))
```
because they both narrow the inferred type `ndarray[Any, dtype[floating[Any]]` and obey the type bounds for shape and dtype.  I agree that they can't be combined, as `Literal[Time(2)]` isn't valid.  It's an either or: you can use literal length axes, or you can use `NewType` to say that the axes mean something specific (e.g `Time`, `Series`).  Functions that require one will reject an array declared as the other unless it is later `cast`.

For me, the original point of the PR/covariance was to use `NewType`, as `int` is rarely going to have any true subclasses.  [The example in PEP 646](https://peps.python.org/pep-0646/#motivation) is pretty close to my use case: a library that handles images and videos, and the need for type checking to handle the difference between a gray movie (with axes height, width, frame) from an RGB image (with axes height, width, channel).  E.g. a function `lower_resolution()` should persist its input type and could apply to both, but a function `convert_gray()` that averages the last axis should only accept an RGB image (or RGB movie).

But I agree with your next comment, that an in-depth example is better in the `numpy.typing` docs, rather than release notes.  How about I shrink this example to just show declarations of `int`, `Literal` and `NewType`?

Reply:
> For me, the original point of the PR/covariance was to use NewType, as int is rarely going to have any true subclasses.

`Literal[N: int]` is a subtype of `int`, and it has many use-caaes, e.g.

```python
type ColumnVector[N: int, T: number[Any]] = ndarray[tuple[Literal[1], N], dtype[T]]
type RotationMatrix[ND: Literal[2, 3], T: floating[Any]] = ndarray[tuple[ND, ND], dtype[T]]
type ImageRGB[W: int, H: int] = ndarray[tuple[H, W, Literal[3]], dtype[uint8]]
```

But that's not all that this PR solves.
NumPy has many examples of functions to accept arrays with either 1 or 2 dimensions,
e.g.:

```python
def f(x: ndarray[tuple[int] | tuple[int, int], dtype[Any]): ...
```

But currently you can't pass `np.ndarray[tuple[int], np.dtype[Any]]`.
If you try do do so, then Pyright reports:

```
Argument of type "ndarray[tuple[int], dtype[Any]]" cannot be assigned to parameter "x" of type "ndarray[tuple[int] | tuple[int, int], dtype[Any]]" in function "g"
  "ndarray[tuple[int], dtype[Any]]" is incompatible with "ndarray[tuple[int] | tuple[int, int], dtype[Any]]"
    Type parameter "_ShapeType@ndarray" is invariant, but "tuple[int]" is not the same as "tuple[int] | tuple[int, int]"
```

This is actually the reason why shape-typing in NumPy is practically impossible at the moment.

---

To my knowledge, there aren't any other issues about these problems specifically.
The way I see it, is that this change allows us to start adding shape-type annotations within NumPy, so this is not a new feature or a bugfix.

Also note that all numpy array constructors currently return `Any` for the shape, and shape-types are almost never used in general.
So unless users currently explicitly `cast` an array to have shape-types, or use a custom shape-typed array constructor or something, this change just isn't *directly* relevant for them.

So I don't think an example is needed.
This is not a feature and not a bugfix; this opens the road to the long-awaited shape-typing of arrays 🎉.

So I don't think an example is really necessary here.
Plus, showing an example of shape-typing in the release notes, even though there's almost no support for it in numpy, is bound to confuse a lot of people.